### ISSUE TITLE
Suggestion: Pluggable storage engines

### ISSUE THREAD

**Issue by elifarley** [rethinkdb/rethinkdb#3966]
This idea is espoused by [Riak] (http://docs.basho.com/riak/latest/ops/building/planning/backends/), [InfluxDB] (http://influxdb.com/docs/v0.8/advanced_topics/sharding_and_storage.html) and probably many other databases.

It would allow RethinkDB users to choose the most appropriate storage backend for their specific needs (think of a storage backend that emphasizes smaller storage footprint versus another emphasizing increased speed).

The current underlying storage engine in RethinkDB could become a project on its own. It could then become an additional storage engine option for other systems, and in turn could get more pull requests in the long run.

In addition to that, a separate project for the current storage engine would decrease the probability of having circular or unnecessary dependencies between the storage engine and the rest of RethinkDB, thus [lowering its maintenance cost] (http://fsharpforfunandprofit.com/posts/cyclic-dependencies/):

> As John Lakos, the author, put it:
*"The maintenance cost of a subsystem can be reduced significantly by avoiding unnecessary dependencies among components"*

**Comment by elifarley** [comment]
I guess that a backend storage engine based on Log-structured merge (LSM) trees like [RocksDB] (http://rocksdb.org/) would help #364, given it has [native support for range-based queries] (https://github.com/facebook/rocksdb/wiki/Prefix-Seek-API-Changes).

Maybe a backend based on [Tokutek's] (http://www.tokutek.com/) [Fractal Tree could help even more] (http://highscalability.com/blog/2014/8/6/tokutek-white-paper-a-comparison-of-log-structured-merge-lsm.html).

**Comment by danielmewes** [comment]
I think our current storage engine is actually really good (it also supports range-based queries btw).

Eventually this will likely happen, but it's not a priority at the moment.


On a side note: John Lakos appears to talk about the costs of maintaining the subsystem. Reducing that doesn't necessarily reduce the cost of maintaining the overall system (the sum of all subsystems if you will). Certain things actually become more difficult if you have less control over some subsystems. That being said keeping subsystems separate is still usually a good idea because it reduces the likelihood of bugs and improves testability.

**Comment by elifarley** [comment]
No doubt about the native storage engine being really good!

But it would be good to switch backend engines and see how the performance profile would change.

Maybe this will grab the attention of some grad student, or maybe I'll find some (substantial) spare time to work on this ;)

**Comment by danielmewes** [comment]
Yes I agree. This would be pretty cool.

If you want to work on this, we'll be more than happy to help you getting started and giving you all the necessary support. I agree that it would likely be a pretty large project.

**Comment by stephanos** [comment]
Yes, it would also be interesting to use an in-memory storage engine for very fast, throw-away databases in integration/functional tests.

**Comment by larkost** [comment]
@stephanos: I hesitate to mention this since it is not really public, and I have to stress that it could go away at any time and is totally unsupported... but we do have a memory-only table for our own debugging use: `rethinkdb._debug_scratch`.

I would never be comfortable using it for anything other than testing (we use it for testing artificial tables without side effects), but it is there for that purpose and it might serve your needs.
